design/recovery-internals.md

```diff
@@ -0,0 +1,142 @@
+# FDB Recovery Internals
+
+FDB uses recovery to handle various failures, such as hardware and network failures. When the current transaction system no longer works properly due to failures, recovery is automatically triggered to create a new generation of the transaction system. 
+
+This document explains at the high level how the recovery works in a single cluster. The audience of this document includes both FDB developers who want to have a basic understanding of the recovery process and database administrators who need to understand why a cluster fails to recover. This document does not discuss the complexity introduced to the recovery process by the multi-region configuration.
+
+
+## Overview
+
+Recovery is triggered on the cluster controller (CC). Coordinators choose a CC through leader election. If the current elected CC does not respond, a new CC will be elected by coordinators. If the majority of coordinators cannot be reached, CC will not be selected, recovery will get stuck.
+
```

Comment:
I didn't know there were multiple cluster controllers :| Maybe worth mentioning for ignorants like me? 

How do the coordinators figure out that they need to start recovery? When they are not able to communicate with certain components I believe. Can we add an example here to clarify?

Reply:
CC can change either because it dies or it cannot reach the majority of coordinators.

Reply:
I see the confusion here based on your question. 

I updated it as `Cluster controller (CC) decides if recovery should be triggered. In case the current running CC crashes or cannot be reached by a majority of coordinators, coordinators will start leader election to select a CC among the stateless process -- the processes which do not have a file behind it, such as the processes that run master. In the rare situation when the majority of coordinators cannot be reached, say a majority of coordinators' machines crash, CC cannot be selected successfully and the recovery will get stuck.`